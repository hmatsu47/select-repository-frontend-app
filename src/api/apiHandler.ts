import { fetchWithTimeout } from "./fetchWithTimeout";

// ベース URI
export const baseUri = "http://10.11.17.43/api";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
