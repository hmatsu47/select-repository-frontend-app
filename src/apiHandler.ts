import { fetchWithTimeout } from "./fetchWithTimeout";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
