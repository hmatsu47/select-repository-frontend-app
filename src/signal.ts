import { createSignal } from "solid-js";
import { AlertColor } from "@suid/material/Alert";

import { ImageItem, RepositoryItem, ServiceItem } from "./type";

export const [service, setService] = createSignal<string | undefined>();
export const [repository, setRepository] = createSignal<string | undefined>();
export const [services, setServices] = createSignal<
  ServiceItem[] | undefined
>();
export const [repositories, setRepositories] = createSignal<
  RepositoryItem[] | undefined
>();
export const [images, setImages] = createSignal<ImageItem[] | undefined>();
export const [imageUri, setImageUri] = createSignal<string | undefined>();
export const [releaseAt, setReleaseAt] = createSignal<Date | undefined>();
export const [lastImageUri, setLastImageUri] = createSignal<
  string | undefined
>();
export const [lastReleasedAt, setLastReleasedAt] = createSignal<
  Date | undefined
>();
export const [isReleased, setIsReleased] = createSignal<boolean | undefined>();
export const [message, setMessage] = createSignal<string | undefined>();
export const [messageSeverity, setMessageSeverity] = createSignal<
  AlertColor | undefined
>();
