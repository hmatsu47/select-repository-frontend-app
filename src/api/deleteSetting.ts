import { ErrorResponse } from "../type";
import { baseUri, deleteApiData } from "./apiHandler";
import {
  imageUri,
  isReleased,
  isReleaseSelected,
  releaseAt,
  service,
  setMessage,
  setMessageSeverity,
  setImageUri,
  setIsReleaseSelected,
  setReleaseAt,
} from "../signal";

export const deleteSetting = async () => {
  if (
    !isReleased() === undefined ||
    !isReleaseSelected() ||
    !imageUri() ||
    !releaseAt()
  ) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: null | ErrorResponse = await deleteApiData(
      `${baseUri}/setting/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setImageUri(undefined);
    setReleaseAt(undefined);
    setIsReleaseSelected(false);
    setMessage("リリースを取り消しました。");
    setMessageSeverity("success");
  };
  void load();
};
