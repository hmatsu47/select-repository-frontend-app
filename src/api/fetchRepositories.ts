import { ErrorResponse, RepositoryItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  service,
  setImages,
  setMessage,
  setMessageSeverity,
  setRepositories,
} from "../signal";

export const fetchRepositories = async () => {
  if (!service()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: RepositoryItem[] | ErrorResponse = await getApiData(
      `${baseUri}/repositories/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setRepositories(undefined);
      setImages(undefined);
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setRepositories(data as RepositoryItem[]);
    setImages(undefined);
    setMessage(undefined);
  };
  void load();
};
