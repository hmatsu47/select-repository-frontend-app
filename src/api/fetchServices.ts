import { ErrorResponse, ServiceItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  setErrorMessage,
  setImages,
  setRepositories,
  setServices,
} from "../signal";

export const fetchServices = async () => {
  const load = async (): Promise<void> => {
    const data: ServiceItem[] | ErrorResponse = await getApiData(
      `${baseUri}/services`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setServices(undefined);
      setRepositories(undefined);
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setServices(data as ServiceItem[]);
    setRepositories(undefined);
    setImages(undefined);
    setErrorMessage(undefined);
  };
  void load();
};
