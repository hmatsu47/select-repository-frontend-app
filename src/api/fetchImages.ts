import { ErrorResponse, ImageItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import { repository, service, setErrorMessage, setImages } from "../signal";

export const fetchImages = async () => {
  if (!service() || !repository()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: ImageItem[] | ErrorResponse = await getApiData(
      `${baseUri}/images/${service()}/${repository()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setImages(data as ImageItem[]);
    setErrorMessage(undefined);
  };
  void load();
};
