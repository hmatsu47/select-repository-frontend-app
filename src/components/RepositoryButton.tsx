import Button from "@suid/material/Button";
import { fetchImages } from "../api/fetchImages";
import { repository, setRepository } from "../signal";
import { RepositoryItem } from "../type";

type Props = {
  repositoryItem: RepositoryItem;
};

export const RepositoryButton = (props: Props) => {
  return (
    <Button
      variant={
        repository() === props.repositoryItem.name ? "contained" : "outlined"
      }
      size="small"
      color={repository() === props.repositoryItem.name ? "primary" : "inherit"}
      onClick={async (e) => {
        setRepository(props.repositoryItem.name);
        await fetchImages();
      }}
      sx={{ textTransform: "none" }}
    >
      {props.repositoryItem.name}
    </Button>
  );
};
