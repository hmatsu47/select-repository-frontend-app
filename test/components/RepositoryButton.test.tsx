// @vitest-environment jsdom

import { describe, expect, test, vi } from "vitest";
import "vi-fetch/setup";
import { mockFetch, mockGet } from "vi-fetch";
import { fireEvent, render } from "@solidjs/testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { ImageItem, RepositoryItem } from "../../src/type";
import { baseUri } from "../../src/api/apiHandler";
import { RepositoryButton } from "../../src/components/RepositoryButton";
import { repository, setService } from "../../src/signal";

describe("<RepositoryButton />", () => {
  beforeEach(() => {
    mockFetch.clearAll();
  });
  afterEach(() => {
    vi.clearAllMocks();
  });
  test("リポジトリボタン", async () => {
    localStorage.removeItem("selectedService");
    localStorage.removeItem("selectedRepository-service1");
    setService("service1");
    const mockImages = mockGet(
      `${baseUri}/images/service1/repository1`
    ).willResolve([
      {
        digest: "",
        pushed_at: new Date(),
        repository_name: "repository1",
        size: 10000,
        uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1@sha256:4c716d4cf211c7b7d2f3233c941771ad0507ea5bacf93b492766aa41ae9f720d",
      } as ImageItem,
    ]);
    const { container, findByText, unmount } = render(() => (
      <RepositoryButton
        repositoryItem={
          {
            name: "repository1",
            uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1",
          } as RepositoryItem
        }
      />
    ));
    // css の名前が動的に変わるので固定値に置換
    const htmlBefore = formatSnapshot(container.innerHTML);
    expect(htmlBefore).toMatchSnapshot();
    // ボタンクリック
    const button = (await findByText("repository1")) as HTMLInputElement;
    expect(button).toHaveTextContent("repository1");
    fireEvent.click(button);
    // ボタンクリック後の照合
    const htmlAfter = formatSnapshot(container.innerHTML);
    expect(htmlAfter).toMatchSnapshot();
    expect(mockImages).toHaveFetched();
    expect(repository()).toBe("repository1");
    unmount();
    localStorage.removeItem("selectedService");
    localStorage.removeItem("selectedRepository-service1");
  });
});
