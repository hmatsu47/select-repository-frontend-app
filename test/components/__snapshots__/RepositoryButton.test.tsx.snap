// Vitest Snapshot v1

exports[`<RepositoryButton /> > リポジトリボタン 1`] = `"<button class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" type=\\"button\\" title=\\"repository1 を選択\\" tabindex=\\"0\\">repository1<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button>"`;

exports[`<RepositoryButton /> > リポジトリボタン 2`] = `"<button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButtonBase-root css-xxxxxx\\" type=\\"button\\" title=\\"repository1 を選択\\" tabindex=\\"0\\">repository1<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button>"`;
